File: frontend/src/lib/api/types/blueprint.ts
/**
 * Blueprint API Types (Stage 3 - Student Blueprint Workflow)
 */

export type BlueprintStatus =
  | 'NOT_STARTED'
  | 'GENERATING'
  | 'GENERATED'
  | 'VALIDATING'
  | 'QA_REJECTED'
  | 'READY_FOR_APPROVAL'
  | 'COMPLETED'
  | 'FAILED'
  | 'APPROVED';

export type BlueprintQAStatus =
  | 'PENDING'
  | 'IN_REVIEW'
  | 'PASSED'
  | 'FAILED';

export type BlueprintJobType =
  | 'FULL_GENERATION'
  | 'SECTION_RETRY';

export type BlueprintJobStatus =
  | 'PENDING'
  | 'RUNNING'
  | 'COMPLETED'
  | 'FAILED';

export type BlueprintSectionKey =
  | 'project_profile'
  | 'tech_stack'
  | 'features'
  | 'specifications'
  | 'mvp'
  | 'duration'
  | 'risks'
  | 'tasks'
  | 'milestones'
  | 'readme';

export const CANONICAL_BLUEPRINT_SECTION_LABELS: Record<BlueprintSectionKey, string> = {
  project_profile: 'Project Profile & Domain Context',
  tech_stack: 'Technical Stack & Architecture',
  features: 'Core Features & System Modules',
  specifications: 'Technical Specifications & Data Models',
  mvp: 'MVP Scope & Validation Criteria',
  duration: 'Timeline & Sprint Duration',
  risks: 'Technical Risks & Mitigations',
  tasks: 'Granular Work Breakdown (Tasks)',
  milestones: 'Stage Milestones & Gate Deliverables',
  readme: 'Production README & Setup Guide',
};

export const CANONICAL_BLUEPRINT_SECTION_ORDER: BlueprintSectionKey[] = [
  'project_profile',
  'tech_stack',
  'features',
  'specifications',
  'mvp',
  'duration',
  'risks',
  'tasks',
  'milestones',
  'readme',
];

export interface BlueprintQAFeedback {
  score: number;
  status: BlueprintQAStatus;
  evaluated_criteria?: Record<string, number>;
  summary?: string;
  issues?: Array<{
    section: string;
    severity: string;
    description: string;
    recommendation: string;
  }>;
  strengths?: string[];
  gaps?: string[];
  recommendations?: string[];
  evaluated_at?: string;
}

export interface BlueprintGenerationProgress {
  completed_sections: BlueprintSectionKey[];
  total_sections: number;
  in_progress_section?: BlueprintSectionKey | null;
  failed_sections?: BlueprintSectionKey[];
}

export interface BlueprintJobSummary {
  id: string;
  job_type: BlueprintJobType;
  status: BlueprintJobStatus;
  error_message?: string | null;
  failed_sections: BlueprintSectionKey[];
  created_at: string;
}

export interface BlueprintStatusResponse {
  blueprint_id: string;
  project_id: string;
  status: BlueprintStatus;
  current_stage: number;
  qa_status: BlueprintQAStatus;
  qa_feedback?: BlueprintQAFeedback | null;
  generation_progress: BlueprintGenerationProgress;
  active_job?: BlueprintJobSummary | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface BlueprintContentResponse {
  project_instance_id?: string;
  blueprint_id?: string;
  project_id?: string;
  status: BlueprintStatus;
  qa_status?: BlueprintQAStatus;
  qa_score?: number | null;
  content: Record<string, unknown>;
  qa_feedback?: BlueprintQAFeedback | null;
}

export interface BlueprintApprovalResponse {
  success?: boolean;
  status: BlueprintStatus | string;
  approved_at: string;
  blueprint_id?: string;
  project_id?: string;
  message?: string;
}

export interface BlueprintRetryRequest {
  target_output_key?: string | null;
}

File: backend/app/api/schemas/blueprint.py
"""
GrowFlow — Blueprint Pydantic Schemas.

Defines typed request and response schemas for:
- /api/v1/projects/{project_id}/blueprint/status
- /api/v1/projects/{project_id}/blueprint/generate
- /api/v1/projects/{project_id}/blueprint/retry
- /api/v1/projects/{project_id}/blueprint/content
- /api/v1/projects/{project_id}/blueprint/approve
"""

from __future__ import annotations

from datetime import datetime
from typing import Any
import uuid

from pydantic import BaseModel, ConfigDict, Field


class BlueprintIssueSchema(BaseModel):
    section: str
    severity: str
    description: str
    recommendation: str


class BlueprintQAFeedbackSchema(BaseModel):
    status: str
    score: int
    summary: str
    evaluated_criteria: dict[str, int] = Field(default_factory=dict)
    issues: list[BlueprintIssueSchema] = Field(default_factory=list)
    recommendations: list[str] = Field(default_factory=list)


class BlueprintGenerationProgressSchema(BaseModel):
    completed_sections: list[str] = Field(default_factory=list)
    total_sections: int = 10
    in_progress_section: str | None = None
    failed_sections: list[str] = Field(default_factory=list)


class BlueprintStatusResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    blueprint_id: uuid.UUID | None = None
    project_instance_id: uuid.UUID
    project_id: uuid.UUID | None = None
    student_id: uuid.UUID
    status: str
    current_step: str | None = None
    progress_percent: int = 0
    current_stage: int = 0
    generation_progress: BlueprintGenerationProgressSchema | None = None
    error_message: str | None = None
    failed_output_key: str | None = None
    qa_status: str
    qa_score: int | None = None
    qa_feedback: dict[str, Any] | None = None
    approved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class BlueprintContentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    project_instance_id: uuid.UUID
    status: str
    qa_status: str
    qa_score: int | None = None
    content: dict[str, Any] = Field(default_factory=dict)


class BlueprintGeneratePayload(BaseModel):
    force_regenerate: bool = False


class BlueprintRetryPayload(BaseModel):
    target_output_key: str | None = None


class BlueprintApproveResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    success: bool
    status: str
    approved_at: datetime
